libs/mbedtls/mbedtls_stubs.c

```diff
@@ -0,0 +1,307 @@
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_ssl_read(value ssl, value buf, value pos, value len) {
+	CAMLparam4(ssl, buf, pos, len);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_read(SSL(ssl), String_val(buf) + Int_val(pos), Int_val(len));
+	CAMLreturn(Val_int(r));
+}
+
+int bio_write_cb(void* ctx, const unsigned char* buf, size_t len) {
+	CAMLparam0();
+	CAMLlocal3(r, s, vctx);
+	vctx = (value)ctx;
+	s = caml_alloc_string(len);
+	memcpy(String_val(s), buf, len);
+	r = caml_callback2(Field(vctx, 1), Field(vctx, 0), s);
+	return Int_val(r);
+}
+
+int bio_read_cb(void* ctx, unsigned char* buf, size_t len) {
+	CAMLparam0();
+	CAMLlocal3(r, s, vctx);
+	vctx = (value)ctx;
+	s = caml_alloc_string(len);
+	r = caml_callback2(Field(vctx, 2), Field(vctx, 0), s);
+	memcpy(buf, String_val(s), len);
+	return Int_val(r);
+}
+
+CAMLprim void ml_mbedtls_ssl_set_bio(value ssl, value p_bio, value f_send, value f_recv) {
+	CAMLparam4(ssl, p_bio, f_send, f_recv);
+	CAMLlocal1(ctx);
+	ctx = caml_alloc(3, 0);
+	Store_field(ctx, 0, p_bio);
+	Store_field(ctx, 1, f_send);
+	Store_field(ctx, 2, f_recv);
+	mbedtls_ssl_set_bio(SSL(ssl), (void*)ctx, bio_write_cb, bio_read_cb, NULL);
+	CAMLreturn0;
+}
+
+CAMLprim value ml_mbedtls_ssl_set_hostname(value ssl, value hostname) {
+	CAMLparam2(ssl, hostname);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_set_hostname(SSL(ssl), String_val(hostname));
+	CAMLreturn(r);
+}
+
+CAMLprim value ml_mbedtls_ssl_setup(value ssl, value conf) {
+	CAMLparam2(ssl, conf);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_setup(SSL(ssl), CFG(conf));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_ssl_write(value ssl, value buf, value pos, value len) {
+	CAMLparam4(ssl, buf, pos, len);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_write(SSL(ssl), String_val(buf) + Int_val(pos), Int_val(len));
+	CAMLreturn(Val_int(r));
+}
+
+// Certificate
+
+CAMLprim value ml_mbedtls_x509_crt_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_x509_crt* cert = malloc(sizeof(mbedtls_x509_crt));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)cert);
+	mbedtls_x509_crt_init(cert);
+	CAMLreturn(obj);
+}
+
+CAMLprim value ml_mbedtls_x509_crt_parse_file(value chain, value path) {
+	CAMLparam2(chain, path);
+	CAMLlocal1(r);
+	r = mbedtls_x509_crt_parse_file(CERT(chain), String_val(path));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_x509_crt_parse_path(value chain, value path) {
+	CAMLparam2(chain, path);
+	CAMLlocal1(r);
+	r = mbedtls_x509_crt_parse_path(CERT(chain), String_val(path));
+	CAMLreturn(Val_int(r));
+}
+
+
+// glue
+
+CAMLprim value hx_cert_load_defaults(value certificate) {
+	CAMLparam0();
+	CAMLlocal1(r);
+	r = 1;
+
+	mbedtls_x509_crt *chain = CERT(certificate);
+
+	#ifdef _WIN32
+	HCERTSTORE store;
+	PCCERT_CONTEXT cert;
+
+	if (store = CertOpenSystemStore(0, "Root")) {
+		cert = NULL;
+		while (cert = CertEnumCertificatesInStore(store, cert)) {
+			r = mbedtls_x509_crt_parse_der(chain, (unsigned char *)cert->pbCertEncoded, cert->cbCertEncoded);
```

Comment:
`Val_int(...)` should be here.

Reply:
We should also terminate this loop as soon as there is an error. Maybe store the result in a regular `int` variable declared before the loop, prepend it to the while condition, update `r` accordingly.

Reply:
I'll have to check the semantics, maybe a failure is acceptable here? I don't know if one failed certificate should bring the entire thing down. But yes just returning the last return value is dodgy for sure.

Reply:
Addressed in https://github.com/HaxeFoundation/haxe/pull/9009/commits/9d7a52c3674bc71152c6ccfd99f21a39bb746420